### PR TITLE
cstring-related cleanup, switch to std::string_view for some cstring API

### PR DESCRIPTION
**Pull request by asl**
- Switch to `std::string_view` for some cstring API

### CODE REVIEW

**Comment by asl** [comment]
@vlstill @fruffy Will you please take a look when you will have a chance? Thanks!

**Comment by fruffy** [comment]
> * Switch to `std::string_view` for some cstring API
> 
> * Get rid of custom implementation of `Util::PathName`. Replace with `std::filesystem::path`, simplify code around

I would consider splitting these into two separate PRs? At least getting rid of Util::PathName, which is a good thing, is fairly involved.

File: lib/path.h [hunk 1]
```diff
@@ -17,57 +17,12 @@ limitations under the License.
...
-    cstring toString() const { return str; }
-    bool isNullOrEmpty() const { return str.isNullOrEmpty(); }
-    bool operator==(const PathName &other) const { return str == other.str; }
-    bool operator!=(const PathName &other) const { return str != other.str; }
-    PathName join(cstring component) const;
-
-    static PathName empty;
-};
+// We used to have local implementation. Make this alias for now
+using PathName = std::filesystem::path;
```
Comment (fruffy) [inline lib/path.h:25]: If we make this a separate PR I'd consider replacing this alias entirely. I do not see a reason to keep it, really.

Or, add a deprecation notice, if possible.

File: control-plane/p4RuntimeSerializer.cpp [hunk 2]
```diff
@@ -1581,6 +1581,7 @@ void P4RuntimeSerializer::serializeP4RuntimeIfRequired(const IR::P4Program *prog
 
 void P4RuntimeSerializer::serializeP4RuntimeIfRequired(const P4RuntimeAPI &p4Runtime,
                                                        const CompilerOptions &options) {
+    // FIXME: get rid of cstring here
```
Comment (fruffy) [inline control-plane/p4RuntimeSerializer.cpp:1584]: What happens if you replace it with `std::filesystem::path`?

Comment (asl) [reply control-plane/p4RuntimeSerializer.cpp:1584]: There are more dependencies here including `parseFileNames` logic, etc. I decided to leave it as-is for now.

File: lib/path.h (hunk 1 above)
Comment (asl) [reply lib/path.h:25]: `std::filesystem::path` is quite long to type, in this sense `PathName` is much shorter even with the namespace. Certainly, this is a bit of bike-shedding :)

But yes, we can get rid of it entirely. Let me split into a separate PR and remove it entirely.

**Comment by asl** [comment]
I submitted https://github.com/p4lang/p4c/pull/4721 for PathName part. Will rebase this one to have just `cstring` bits afterwards

File: lib/cstring.h [hunk 3]
```diff
@@ -254,17 +254,11 @@ class cstring {
     }
     cstring substr(size_t start, size_t length) const;
     cstring replace(char find, char replace) const;
-    cstring replace(cstring find, cstring replace) const;
+    cstring replace(std::string_view find, std::string_view replace) const;
```
Comment (fruffy) [inline lib/cstring.h:257]: This might be the only breaking change in this PR, right?

**Review by fruffy — COMMENTED** [review]
This ready? I guess the only thing that needs to change is PR title and description.

File: lib/cstring.h (hunk 3 above)
Comment (asl) [reply lib/cstring.h:257]: Yes. Though if before explicit-cstring-construction it was like `replace("foo", "bar")`, then after this PR it will start working again w/o changes :)

**Review by ChrisDodd — APPROVED** [review]

File: lib/cstring.h (hunk 3 above)
Comment (vlstill) [reply lib/cstring.h:257]: Hmm, I wonder why not return a `std::string` here. In many cases, we would not need the result to be `cstring` and if the caller needs it, they can always do that explicitly.

Comment (asl) [reply lib/cstring.h:257]: I checked this case, and there is a mixture of usages. So I decided to re-evaluate this when we'll disable implicit conversion to / from std::string – then I will catch all such places :)
